azkaban-common/src/main/java/azkaban/executor/mail/TemplateBasedMailCreator.java

```diff
@@ -0,0 +1,211 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package azkaban.executor.mail;
+
+import azkaban.executor.ExecutableFlow;
+import azkaban.executor.ExecutionOptions;
+import azkaban.executor.ExecutionOptions.FailureAction;
+import azkaban.utils.EmailMessage;
+import azkaban.utils.Emailer;
+import azkaban.utils.Utils;
+import com.google.common.base.Charsets;
+import org.apache.commons.io.IOUtils;
+import org.apache.log4j.Logger;
+
+import java.io.InputStream;
+import java.nio.charset.Charset;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
+import java.util.*;
+import java.util.function.BiFunction;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class TemplateBasedMailCreator implements MailCreator {
+
+  private static final Logger logger = Logger.getLogger(TemplateBasedMailCreator.class);
+
+  public static final String NAME = "template-based";
+  private static final DateFormat DATE_FORMATTER = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss z");
+
+  private static String convertMSToString(final long timeInMS) {
+    return timeInMS < 0 ? "N/A" : DATE_FORMATTER.format(new Date(timeInMS));
+  }
+
+  private static String readResource(final String file) throws Exception {
+    final InputStream is = TemplateBasedMailCreator.class.getResourceAsStream(file);
+    return IOUtils.toString(is, Charsets.UTF_8).trim();
+  }
+
+  private static String readFile(final String file) throws Exception {
+    return new String(Files.readAllBytes(Paths.get(file)), Charset.defaultCharset());
+  }
+
+  private String errorTemplate;
+  private String firstErrorTemplate;
+  private String successTemplate;
+
+  public static TemplateBasedMailCreator fromResources(String emailTemplatesPath) throws Exception {
+    return new TemplateBasedMailCreator(
+      readResource(emailTemplatesPath + "/errorEmail.html"),
+      readResource(emailTemplatesPath + "/firstErrorEmail.html"),
+      readResource(emailTemplatesPath + "/successEmail.html")
+    );
+  }
+
+  public static TemplateBasedMailCreator fromPath(String emailTemplatesPath) throws Exception {
+    return new TemplateBasedMailCreator(
+      readFile(emailTemplatesPath + "/errorEmail.html"),
+      readFile(emailTemplatesPath + "/firstErrorEmail.html"),
+      readFile(emailTemplatesPath + "/successEmail.html")
+    );
+  }
+
+  public TemplateBasedMailCreator(String errorTemplate, String firstErrorTemplate, String successTemplate) throws Exception {
+    logger.info("TemplateBasedMailCreator(errorTemplate=" + errorTemplate + ", firstErrorTemplate=" + firstErrorTemplate + ", successTemplate=" + successTemplate + ")");
```

Comment:
Do you think it's necessary to print out all the templates in the server log? We can just view them in .html files, right?

Reply:
that's a fair point, fixing...